File: golang-be/Encounters/service/UserExperienceService.go
package service

import (
	"encounters.xws.com/model"
	"encounters.xws.com/repo"
)

type UserExperienceService struct {
	UserExperienceRepo *repo.UserExperienceRepository
}

func NewUserExperienceService(repo *repo.UserExperienceRepository) *UserExperienceService {
    return &UserExperienceService{
        UserExperienceRepo: repo,
    }
}

func (service *UserExperienceService) Create(userExperience *model.UserExperience) error {
	err := service.UserExperienceRepo.Create(userExperience)
	if err != nil {
		return err
	}
	return nil
}

func (service *UserExperienceService) Delete(id int) error {
	userExperience, err := service.UserExperienceRepo.Get(id)
	if err != nil{
		return err
	}
	err1 := service.UserExperienceRepo.Delete(&userExperience)
	if err1 != nil {
		return err1
	}
	return nil
}

func (service *UserExperienceService) AddXP(id int, xp int) (model.UserExperience, error) {
	userExperience, err := service.UserExperienceRepo.Get(id)
	if err != nil{
		return userExperience, err
	}
	userExperience.XP += xp
	userExperience.Level = model.CalculateLevel(&userExperience)
	dbResult := service.UserExperienceRepo.Update(&userExperience)
	if dbResult.Error != nil {
		return userExperience, dbResult
	}
	return userExperience, nil
}

func (service *UserExperienceService) GetByUserId(id int) (model.UserExperience, error) {
	userExperience, err := service.UserExperienceRepo.GetByUserId(id)
	if err != nil {
		return userExperience, 
		err
	}
	return userExperience, nil
}

File: tourism-be/src/Modules/Encounters/Explorer.Encounters.Core/UseCases/UserExperienceService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using AutoMapper;
using Explorer.BuildingBlocks.Core.UseCases;
using Explorer.Encounters.API.Dtos;
using Explorer.Encounters.API.Public;
using Explorer.Encounters.Core.Domain;
using Explorer.Encounters.Core.Domain.RepositoryInterfaces;
using FluentResults;

namespace Explorer.Encounters.Core.UseCases
{
    public class UserExperienceService : CrudService<UserExperienceDto, UserExperience>, IUserExperienceService
    {
        IUserExperienceRepository _userExperienceRepository;
        public UserExperienceService(ICrudRepository<UserExperience> crudRepository, IUserExperienceRepository userExperienceRepository, IMapper mapper) : base(crudRepository, mapper)
        {
            _userExperienceRepository = userExperienceRepository;
        }

        public Result<UserExperienceDto> AddXP(long id,int xp)
        {
            var userExperience = CrudRepository.Get(id);
            userExperience.AddXP(xp);
            userExperience.CalculateLevel();
            CrudRepository.Update(userExperience);
            UserExperienceDto dto = new UserExperienceDto();
            dto.UserId = userExperience.UserId;
            dto.XP = userExperience.XP;
            dto.Id = (int)userExperience.Id;
            dto.Level = userExperience.Level;
            return dto;

        }

        public Result<UserExperienceDto> GetByUserId(long userId)
        {
            var userExperience = _userExperienceRepository.GetByUserId(userId);
            UserExperienceDto dto = new UserExperienceDto();
            dto.UserId = userExperience.UserId;
            dto.XP = userExperience.XP;
            dto.Id = (int)userExperience.Id;
            dto.Level = userExperience.Level;
            return dto;
        }
    }
}
